cdap-ui/app/features/datasets/controllers/tabs/query-model.js

```diff
@@ -0,0 +1,79 @@
+angular.module(PKG.name+'.feature.datasets')
+  .factory('QueryModel', function ($q, mySettings) {
+
+      function QueryModel(dataSrc, key, data) {
+        this.dataSrc = dataSrc;
+        this.key = key;
+        this.data = data || [];
+        if (this.data.length) {
+          this.sync();
+        }
+      }
+
+      QueryModel.prototype.add = function (entry) {
+        if(this.data.indexOf(entry) == -1) {
+          this.data.push(entry);
+          this.sync();
+        }
+      };
+
+      QueryModel.prototype.set = function (entries) {
+        if(angular.isArray(entries)) {
+          this.data = entries;
+          this.sync();
+        }
+      };
+
+      QueryModel.prototype.get = function () {
+        var self = this;
+        var deferred = $q.defer();
+        var requests = [];
+        requests.push(mySettings.get(this.key));
+        requests.push(this.dataSrc.request({
+          _cdapNsPath: '/data/explore/queries',
+          })
+        );
+        $q.all(requests).then(function (responses) {
+          self.data = self.mergeQueries(responses[0], responses[1]);
+          deferred.resolve(self.data);
+          self.sync();
+        });
+        return deferred.promise;
+      };
+
+      /**
+       * Merges queries stored in user settings and queries pulled from server
+       * using query_handle.
+       */
+      QueryModel.prototype.mergeQueries = function (userQueries, apiQueries) {
```

Comment:
if they are expired, should we delete them, then it would not make sense to store them in user settings. The table also allows old queries to be clicked and previewed. I've attached a sample response from the server regarding queries, should I check is active and whether it is contained in this list?
[
{
timestamp: 1427414764599,
statement: "SELECT * FROM dataset_wordcounts LIMIT 5",
status: "FINISHED",
query_handle: "71339d45-da97-4557-a3ef-95c94a4f2519",
has_results: true,
is_active: true
}
]